File: ios/TerminalDeck/Transport/Heartbeat.swift
/**
 * One timer, every socket.
 *
 * ## Why the ping cannot be dropped
 *
 * A long-lived TCP connection through a carrier NAT dies silently when the
 * mapping is reclaimed, which happens after 30–60 seconds of silence on most
 * mobile networks; TCP keepalive defaults to hours and does not save it. So a
 * phone holding a socket open to a machine has to say something regularly or it
 * will find out the socket is dead only when somebody types into it. That is the
 * `7.9` exception: this timer is load-bearing, and dropping it does not save
 * work, it loses sessions.
 *
 * ## Why there is exactly one of them
 *
 * `7.10`, and on a phone it is not a style point. Every host this phone is paired
 * with holds a socket, and every socket needs that ping — so N hosts on N
 * independent timers is N chances per cycle to wake the radio out of its idle
 * state. The cellular radio is the expensive thing here, not the CPU: a wake-up
 * costs the same whether one byte or five hundred follow it, and it keeps the
 * radio in a high-power state for seconds afterwards. Three unsynchronised
 * 25-second timers therefore cost roughly three times what three synchronised
 * ones do, for identical traffic.
 *
 * Folding them into one tick makes the whole set cost one wake-up: the pings for
 * every host go out in the same turn of the run loop, into the same radio window,
 * and the answers come back into it too.
 *
 * ## Measured
 *
 * By `scripts/keepalive-cost.ts`, which runs the shipping Noise channel, relay
 * envelope and frame writers rather than adding up remembered constants — an
 * earlier version of this comment did the latter and got two of the four numbers
 * wrong. One tick every 25 s, so 144 ticks an hour whatever the number of hosts.
 * A `ping` and a `pong` are both 12 bytes of JSON, which the sealed channel's
 * tag, the version byte, the 17-byte envelope header and the WebSocket frame
 * carry to **52 bytes out** (a client masks, which costs four) and **48 bytes
 * in** — 100 bytes per host per tick. Three paired machines is therefore
 * **43.2 kB an hour** of keepalive across all of them, in **144 radio windows
 * rather than 432**. The full reading is in `docs/multi-host-battery.md`.
 */

import Foundation

@MainActor
final class Heartbeat {

    /**
     * The app's single ping cadence.
     *
     * Well under the 30 seconds after which idle NAT and proxy entries start
     * being reclaimed, so no socket is ever idle long enough to be collected.
     */
    static let interval: TimeInterval = 25

    /// The one every transport joins. A second instance would be a second timer,
    /// which is the thing this type exists to prevent.
    static let shared = Heartbeat()

    private var members: [ObjectIdentifier: () -> Void] = [:]
    private var ticking = false
    /// Bumped when the set empties, so a loop from a previous run stops rather
    /// than racing a new one into existence.
    private var generation = 0

    /// Only the tests read these, to prove there is one loop and that it stops.
    var memberCount: Int { members.count }
    var isTicking: Bool { ticking }

    init() {}

    /**
     * Beat with everyone else.
     *
     * The owner is held by identity and the closure weakly by the caller's own
     * capture list — nothing here retains a transport, because a transport that
     * is deallocated without leaving would otherwise be kept alive by the timer
     * that exists to check on it.
     */
    func join(_ owner: AnyObject, beat: @escaping () -> Void) {
        members[ObjectIdentifier(owner)] = beat
        start()
    }

    func leave(_ owner: AnyObject) {
        members.removeValue(forKey: ObjectIdentifier(owner))
        // Stopped rather than left spinning over an empty set. A timer with no
        // consumer is pure cost, and on a phone it is cost that shows up as
        // battery in somebody's settings screen with this app's name on it.
        if members.isEmpty { stop() }
    }

    /**
     * The app came to the foreground, or the network changed.
     *
     * **Beats now, then realigns.** The earlier version of this deliberately did
     * not — the reasoning was that firing every host at once on resume was a
     * thundering herd worth avoiding, so the first beat was left a full interval
     * away. That reasoning was exactly backwards, and it cost the app the one
     * property it must never lose.
     *
     * Coming back from suspension is the moment a socket is *most* likely to be
     * dead: the app has not run for minutes or hours, and nothing tells it that a
     * carrier NAT reclaimed the mapping while it was away. Pushing the first
     * check a full interval out meant the badge could read Connected for 25
     * seconds plus the 10-second grace against nothing at all — and it was
     * observed doing so.
     *
     * The herd was never the problem either: firing every host in the same turn
     * of the run loop is *the entire point* of a shared tick. One radio window,
     * N pings.
     */
    func realign() {
        guard ticking else { return }
        stop()
        for beat in Array(members.values) { beat() }
        start()
    }

    private func start() {
        guard !ticking, !members.isEmpty else { return }
        ticking = true
        generation &+= 1
        let epoch = generation
        Task { @MainActor [weak self] in
            while true {
                try? await Task.sleep(for: .seconds(Heartbeat.interval))
                guard let self, self.ticking, self.generation == epoch else { return }
                // Copied before iterating: a beat may leave the set — a transport
                // that finds its socket dead tears itself down from inside this
                // loop — and mutating a dictionary while walking it is a crash,
                // not a warning.
                for beat in Array(self.members.values) { beat() }
            }
        }
    }

    private func stop() {
        ticking = false
        generation &+= 1
    }
}

File: android/app/src/main/java/dev/terminaldeck/android/transport/Heartbeat.kt
package dev.terminaldeck.android.transport

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * One timer, every socket.
 *
 * ## Why the ping cannot simply be dropped
 *
 * A long-lived TCP connection through a carrier NAT dies silently when the mapping is reclaimed,
 * which happens after 30–60 seconds of silence on most mobile networks; TCP keepalive defaults to
 * hours and does not save it. A phone holding a socket open to a machine therefore has to say
 * something regularly or it will discover the socket is dead only when somebody types into it — and
 * "somebody types into it" is the one moment this app must not be wrong about. So the poll is
 * load-bearing rather than laziness: dropping it does not save work, it loses sessions.
 *
 * ## Why there is exactly one of them
 *
 * Every paired machine holds its own socket and every socket needs that ping, so N machines on N
 * independent timers is N chances per cycle to wake the radio out of its idle state. The radio is
 * the expensive part, not the CPU: a wake-up costs about the same whether one byte or five hundred
 * follow it, and it holds the radio in a high-power state for seconds afterwards. Three
 * unsynchronised 25-second timers cost roughly three times what three synchronised ones do, for
 * identical traffic.
 *
 * Folding them into one tick makes the whole set cost one wake-up: every host's ping goes out in the
 * same turn of the loop, into the same radio window, and the answers come back into it.
 *
 * ## The cadence, and what it adds up to
 *
 * One tick every **25 seconds** — comfortably under the 30 after which idle NAT and proxy entries
 * start being reclaimed — so 144 ticks an hour whatever the number of machines. A `ping` is 12 bytes
 * of JSON and a `pong` 14, which the sealed envelope and the WebSocket frame carry to roughly 60 and
 * 62 on the wire. Three paired machines is therefore about 53 KB an hour of keepalive across all of
 * them, in 144 radio windows rather than 432.
 *
 * ## Two phases, still one timer
 *
 * The tick sends, waits out the grace, and then asks every member whether it was answered. That
 * second pass wakes the CPU and sends nothing, so it costs no radio window at all — which is why it
 * is worth keeping the honest 10-second deadline rather than folding the check into the next tick
 * and taking a minute to notice a dead socket.
 */
class Heartbeat(
    /**
     * Where the loop runs. Defaulted to a scope of this class's own rather than taken from a
     * caller: the tick outlives any one transport, and a loop parented to the first transport's
     * scope would stop the moment that machine was forgotten, silently taking the keepalive for
     * every other machine with it.
     */
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default),
    private val intervalMs: Long = INTERVAL_MS,
    private val graceMs: Long = GRACE_MS,
) {

    /**
     * What a socket does when the tick comes round.
     *
     * Two calls rather than one, because a keepalive is a question and an answer and only the
     * member knows whether its answer arrived.
     */
    interface Member {

        /**
         * Say something on this socket.
         *
         * False means there is nothing to say it on any more, and the member is dropped from the
         * tick rather than asked again — a transport that has torn its socket down must not keep a
         * timer alive on the strength of having once joined one.
         */
        fun ping(): Boolean

        /** The grace has run out. Answered or not, the member decides what that means. */
        fun pongDue()
    }

    private val lock = Any()
    private val members = LinkedHashMap<Any, Member>()
    private var loop: Job? = null

    /** Only the tests read these, to prove there is one loop and that it stops. */
    val memberCount: Int get() = synchronized(lock) { members.size }

    val isTicking: Boolean get() = synchronized(lock) { loop?.isActive == true }

    /**
     * Beat with everyone else.
     *
     * Keyed by the owner rather than by the member, so a transport that reconnects replaces its own
     * entry instead of accumulating one per connection — which would have every socket pinged as
     * many times per tick as it had ever been opened.
     */
    fun join(owner: Any, member: Member) {
        synchronized(lock) {
            members[owner] = member
            start()
        }
    }

    fun leave(owner: Any) {
        synchronized(lock) {
            members.remove(owner)
            // Stopped rather than left spinning over an empty set. A timer with nothing to do is
            // pure cost, and on a phone it is cost that shows up in a battery screen with this
            // app's name against it.
            if (members.isEmpty()) stop()
        }
    }

    /**
     * The app came back to the foreground, or the network changed.
     *
     * The tick is realigned rather than left where it was, so the first beat after a resume is a
     * full interval away instead of every machine being pinged at once on top of the reconnects
     * that are already in flight.
     */
    fun realign() {
        synchronized(lock) {
            if (loop?.isActive != true) return
            stop()
            start()
        }
    }

    /** Caller holds [lock]. */
    private fun start() {
        if (loop?.isActive == true || members.isEmpty()) return
        loop = scope.launch {
            while (isActive) {
                delay(intervalMs - graceMs)
                // Snapshotted, and the beats run outside the lock. A member that finds its socket
                // dead tears itself down from inside `ping`, which calls back into `leave` — under
                // the lock that would be a deadlock, and iterating the live map would be a
                // concurrent modification.
                for ((owner, member) in snapshot()) {
                    if (!member.ping()) leave(owner)
                }
                delay(graceMs)
                for ((_, member) in snapshot()) member.pongDue()
            }
        }
    }

    /** Caller holds [lock]. */
    private fun stop() {
        loop?.cancel()
        loop = null
    }

    private fun snapshot(): List<Pair<Any, Member>> =
        synchronized(lock) { members.entries.map { it.key to it.value } }

    companion object {

        /** Well under the 30 seconds after which idle NAT entries start being reclaimed. */
        const val INTERVAL_MS = 25_000L

        /** How long a machine has to answer before its socket is treated as gone. */
        const val GRACE_MS = 10_000L

        /**
         * The one every transport joins.
         *
         * A second instance would be a second timer, which is the thing this class exists to
         * prevent — so the app has exactly one and the tests build their own over a test scheduler.
         */
        val shared = Heartbeat()
    }
}
